File: 0610/HyowonPark/표편집.js
/**
 *
 * @param {행} n
 * @param {시작 행} k
 * @param {명령어 문자열 배열} cmd
 * @returns {삭제된 행은 X, 그대로면 O 문자열}
 */

function solution(n, k, cmd) {
  let head = null;
  let tail = null;
  let selected = null;
  let deletedStack = [];

  for (let i = 0; i < n; i++) {
    const newNode = {
      i,
      originalIdx: i,
      next: null,
      prev: null,
    };

    if (!head) {
      head = newNode;
      tail = newNode;
    } else {
      tail.next = newNode;
      newNode.prev = tail;
      tail = newNode;
    }

    if (i === k) {
      selected = newNode;
    }
  }

  for (let i = 0; i < cmd.length; i++) {
    let [action, amount] = cmd[i].split(" ");
    amount = +amount;

    switch (action) {
      case "D":
        for (let j = 0; j < amount; j++) {
          if (selected.next) {
            selected = selected.next;
          }
        }
        break;

      case "U":
        for (let j = 0; j < amount; j++) {
          if (selected.prev) {
            selected = selected.prev;
          }
        }
        break;
      case "C":
        deletedStack.push(selected);

        if (selected.prev) {
          selected.prev.next = selected.next;
        } else {
          head = selected.next;
        }

        if (selected.next) {
          selected.next.prev = selected.prev;
        } else {
          tail = selected.prev;
        }

        if (selected.next) {
          selected = selected.next;
        } else {
          selected = selected.prev;
        }
        break;
      case "Z":
        const restored = deletedStack.pop();

        if (restored.prev) {
          restored.prev.next = restored;
        } else {
          head = restored;
        }

        if (restored.next) {
          restored.next.prev = restored;
        } else {
          tail = restored;
        }
        break;
    }
  }

  let result = new Array(n).fill("X");
  let current = head;
  while (current) {
    result[current.originalIdx] = "O";
    current = current.next;
  }

  return result.join("");
}

const r1 = solution(8, 2, [
  "D 2",
  "C",
  "U 3",
  "C",
  "D 4",
  "C",
  "U 2",
  "Z",
  "Z",
  "U 1",
  "C",
]);
console.log(r1);

File: 0610/SeoyeonHong/표 편집.py
# https://school.programmers.co.kr/learn/courses/30/lessons/81303
# 모든 명령어를 수행한 후 표의 상태와 처음 주어진 표의 상태를 비교
def solution(n, k, cmd):
    answer = ['O'] * n
    prev = list(range(-1, n-1))
    next = list(range(1, n+1))
    cur = k
    deleted = []
    for c in cmd:
        s = c.split()
        if s[0] == 'U':
            for _ in range(int(s[1])):
                cur = prev[cur]
        elif s[0] == 'D':
            for _ in range(int(s[1])):
                cur = next[cur]
        elif s[0] == 'C':     
            deleted.append(cur)
            if prev[cur] != -1: # 마지막 요소 제거할 경우
                next[prev[cur]] = next[cur]
            if next[cur] != n:# 첫번째 요소 제거할 경우
                prev[next[cur]] = prev[cur]
            
            cur = next[cur] if next[cur] != n else prev[cur]
        else: # 되돌리기
            d = deleted.pop()
            if prev[d] != -1:
                next[prev[d]] = d
            if next[d] != n:
                prev[next[d]] = d

    for d in deleted:
        answer[d] = 'X'
    
    return ''.join(answer)